exonum/src/blockchain/block.rs

```diff
@@ -207,6 +210,103 @@ impl BlockProof {
             non_exhaustive: (),
         }
     }
+
+    /// Verifies that the block in this proof is endorsed by the Byzantine majority of provided
+    /// validators.
+    pub fn verify(&self, validator_keys: &[PublicKey]) -> Result<(), ProofError> {
+        if self.precommits.len() < byzantine_quorum(validator_keys.len()) {
+            return Err(ProofError::NoQuorum);
+        }
+        if self.precommits.len() > validator_keys.len() {
+            return Err(ProofError::DoubleEndorsement);
+        }
+
+        let correct_heights = self
+            .precommits
+            .iter()
+            .all(|precommit| precommit.payload().height == self.block.height);
+        if !correct_heights {
+            return Err(ProofError::IncorrectHeight);
+        }
+
+        let block_hash = self.block.object_hash();
+        let correct_block_hashes = self
+            .precommits
+            .iter()
+            .all(|precommit| precommit.payload().block_hash == block_hash);
+        if !correct_block_hashes {
+            return Err(ProofError::IncorrectBlockHash);
+        }
+
+        let mut endorsements = vec![false; validator_keys.len()];
+        for precommit in &self.precommits {
+            let validator_id = precommit.payload().validator.0 as usize;
+            let expected_key = *validator_keys
+                .get(validator_id)
+                .ok_or(ProofError::IncorrectValidatorId)?;
+            if expected_key != precommit.author() {
+                return Err(ProofError::ValidatorKeyMismatch);
+            }
+            if endorsements[validator_id] {
+                return Err(ProofError::DoubleEndorsement);
+            }
+            endorsements[validator_id] = true;
+        }
+        debug_assert!(
```

Comment:
Don't we really need it assert in the release code ?

Reply:
This assertion should always hold (otherwise, it would be a check, not an assertion). Indeed, we've checked that there are +2/3 precommits and that there are no double endorsements; hence, the block should be approved by +2/3 validators. The `debug_assert` is there to better illustrate the check logic.